More template classes derived from `SPOSetT<T>`

ylee88 opened the pull request:
## Proposed changes

Introducing templates for the classes derived from `SPOSetT<T>`. This is the following change towards a unified build for real, complex, and mixed-precision, started from #5306, and separated from #5355.

## What type(s) of changes does this code introduce?

- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Linux with GCC

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

ye-luo commented:
Test this please

ye-luo commented:
Test this please

ye-luo commented on src/QMCWaveFunctions/tests/ConstantSPOSet.cpp:
```diff
@@ -13,81 +13,99 @@
+  this->OrbitalSetSize = norbitals;
+  ref_psi_.resize(numparticles_, this->OrbitalSetSize);
+  ref_egrad_.resize(numparticles_, this->OrbitalSetSize);
+  ref_elapl_.resize(numparticles_, this->OrbitalSetSize);
```
Could you replace all the `this->OrbitalSetSize` with `SPOSet::OrbitalSetSize`

ylee88 replied:
Addressed in 596c21a3aa29f2b71c0457d15f3e7e62fc98656b.

ye-luo commented on src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:
```diff
@@ -947,8 +963,10 @@ TEST_CASE("RotatedSPOs mw_ APIs", "[wavefunction]")
     //just a loop over the single walker API.
-    RotatedSPOs rot_spo0("rotated0", std::make_unique<DummySPOSetWithoutMW>("no mw 0"));
-    RotatedSPOs rot_spo1("rotated1", std::make_unique<DummySPOSetWithoutMW>("no mw 1"));
+    using ValueType = SPOSet::ValueType;
```
Can we do `Value` here?

ylee88 replied:
Addressed in 60101d20b19fd4d1e476f2e4b201782bacb04250.

ye-luo commented on src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:
```diff
@@ -1005,9 +1023,10 @@ TEST_CASE("RotatedSPOs mw_ APIs", "[wavefunction]")
     //in the underlying SPO and not using the default SPOSet implementation which
     //loops over single walker APIs (which have different values enforced in
     // DummySPOSetWithoutMW
+    using ValueType = SPOSet::ValueType;
```
Can we do `Value` here?

ylee88 replied:
Addressed in 60101d20b19fd4d1e476f2e4b201782bacb04250.

ylee88 commented:
I have changed `ValueType`s to `Value` only the lines that I introduced in this PR. If you want to delete every `*Type` (e.g., `RealType`, `ComplexType`) in `test_RotatedSPOs.cpp` (or other files), let me know.

ye-luo commented:
> I have changed `ValueType`s to `Value` only the lines that I introduced in this PR. If you want to delete every `*Type` (e.g., `RealType`, `ComplexType`) in `test_RotatedSPOs.cpp` (or other files), let me know.

We can tackle `*Type` in `test_RotatedSPOs` when we update `RotatedSPOs`

ye-luo reviewed (approved):
Still need non-ANL approval.

prckent reviewed (approved):
Thanks Youngjun

prckent commented:
Test this please
